cudax/include/cuda/experimental/__memory_resource/any_resource.cuh

```diff
@@ -42,17 +42,150 @@
+    }
+  }
+
+  _Shared_resource& operator=(const _Shared_resource& __other) noexcept
+  {
+    if (this != &__other)
+    {
+      _Shared_resource(__other).swap(*this);
+    }
+
+    return *this;
+  }
+
+  _Shared_resource& operator=(_Shared_resource&& __other) noexcept
+  {
+    if (this != &__other)
+    {
+      _Shared_resource(_CUDA_VSTD::move(__other)).swap(*this);
+    }
+
+    return *this;
+  }
+
+  void swap(_Shared_resource& __other) noexcept
+  {
+    _CUDA_VSTD::swap(__control_block, __other.__control_block);
+  }
+
+  _CCCL_NODISCARD void* allocate(size_t __size, size_t __alignment)
+  {
+    return __control_block->__resource.allocate(__size, __alignment);
+  }
+
+  void deallocate(void* __ptr, size_t __size, size_t __alignment)
+  {
+    __control_block->__resource.deallocate(__ptr, __size, __alignment);
+  }
+
+  friend void swap(_Shared_resource& __left, _Shared_resource& __right) noexcept
+  {
+    __left.swap(__right);
+  }
+
+  _CCCL_NODISCARD_FRIEND bool operator==(const _Shared_resource& __left, const _Shared_resource& __right)
+  {
+    if (__left.__control_block == __right.__control_block)
+    {
+      return true;
+    }
+
+    if (__left.__control_block == nullptr || __right.__control_block == nullptr)
+    {
+      return false;
+    }
+
+    return __left.__control_block->__resource == __right.__control_block->__resource;
+  }
+
+  _CCCL_NODISCARD_FRIEND bool operator!=(const _Shared_resource& __left, const _Shared_resource& __right)
+  {
+    return !(__left == __right);
+  }
+
+  _LIBCUDACXX_TEMPLATE(class _Property)
+  _LIBCUDACXX_REQUIRES(::cuda::has_property<_Resource, _Property>)
+  _CCCL_NODISCARD_FRIEND __property_value_t<_Property> get_property(const _Shared_resource& __self, _Property) noexcept
+  {
+    return get_property(__self.__control_block->__resource, _Property{});
+  }
+
+private:
+  friend struct _Shared_resource<_Resource, _CUDA_VMR::_AllocType::_Async>;
+
+  struct _Control_block
+  {
+    _Resource __resource;
+    _CUDA_VSTD::atomic<int> __ref_count;
+  };
+
+  _Control_block* __control_block;
+};
+
+template <class _Resource>
+struct _Shared_resource<_Resource, _CUDA_VMR::_AllocType::_Async>
+    : _Shared_resource<_Resource, _CUDA_VMR::_AllocType::_Default>
+{
+  using _Shared_resource<_Resource, _CUDA_VMR::_AllocType::_Default>::_Shared_resource;
+
+  _CCCL_NODISCARD void* async_allocate(size_t __size, size_t __alignment, ::cuda::stream_ref __stream)
+  {
+    return this->__control_block->__resource.async_allocate(__size, __alignment, __stream);
+  }
+
+  void async_deallocate(void* __ptr, size_t __size, size_t __alignment, ::cuda::stream_ref __stream)
+  {
+    this->__control_block->__resource.async_deallocate(__ptr, __size, __alignment, __stream);
+  }
+};
+
+template <class _Resource, class __resource_t = _CUDA_VSTD::remove_cvref_t<_Resource>>
+using __as_small_resource_t =
+  _CUDA_VSTD::_If<_CUDA_VMR::_IsSmall<__resource_t>(), __resource_t, _Shared_resource<__resource_t>>;
```

Comment:
Question: Why not using a `shared_ptr` from the host standard library if this is host only anyway?

If we decide that we want to implement it by hand, then we should definitely pull this into a separate header